### ISSUE TITLE
NuGet package license questions

### ISSUE THREAD

**Issue by CodeReaper** [microsoft/vstest#3871]
The license associated with the NuGet package for Microsoft.NET.Test.Sdk is the Microsoft .NET Library license even though the repository appears to be licensed under MIT.

I realise this has been brought up before in [#1691](https://github.com/microsoft/vstest/issues/1691), but it is a little light on actual help. It left me with two questions:

1. Are the packages built in this repository not meant to be used in open source development?
2. Is there a recommended (and legal) way of using this package in an open source project that is compliant with the MSL license?

**Comment by nohwnd** [comment]
@Haplois I think you know the most about this.

**Comment by davidaramantSEP** [comment]
I am currently doing a license audit on my project and I'm also confused about this... Is the NuGet package dual licensed, or is it actually using a different license than the source code?

**Comment by rouke-broersma** [comment]
@CodeReaper which parts of the license do you think make it illegal to use these libraries in open source?

**Comment by CodeReaper** [comment]
> @CodeReaper which parts of the license do you think make it illegal to use these libraries in open source?

@rouke-broersma

I am no lawyer, but my concern began with the fact that repository is licensed with MIT and the NuGet package lists a different license.

I then assumed this license to be incompatible with open source since I am unable to look it up at:
- https://opensource.org/licenses/alphabetical
- https://spdx.org/licenses/
- https://tldrlegal.com

The NuGet package has had its current license for almost 5 years and the above sites only list MS-PL and MS-RL as valid open source Microsoft licenses.

Therefore I do know not if these libraries are compatible with open source or not. Since I cannot say Yes, I am defaulting to No.

**Comment by rouke-broersma** [comment]
I believe it's this license: https://www.microsoft.com/web/webpi/eula/aspnetcomponent_rtw_enu.htm

Just because it's not on tldr legal and such doesn't mean it's not a compatible license.

I'm also not a legal expert and would love someone from Microsoft to weigh in but from just reading the license terms it doesn't seem to block using nor redistribution the packages. It does seem to restrict modifying the nuget packages or reverse engineering them.

They seem to want to protect their compiled packages with this license for some reason, but the source code itself is MIT licensed. So don't reverse and modify the nuget package and compiled dlls but rather take the actual sources and compile and package it yourself if you wish to modify the sources?

Again no legal expert and ms should clarify but this is what I gather from the licenses and the difference in licenses.

**Comment by Evangelink** [comment]
We are looking into it and will come back with an official answer.

**Comment by martinsuchan** [comment]
I'd like to refresh this issue - it's really quite strange situation:
The [MSTest git repo ](https://github.com/microsoft/testfx) has MIT license and the [MSTest.TestFramework NuGet package](https://www.nuget.org/packages/MSTest.TestFramework) has MIT license as well.
The [VSTest git repo](https://github.com/microsoft/vstest), this one, has MIT license, BUT the published [Microsoft.Net.Test.Sdk NuGet package](https://www.nuget.org/packages/Microsoft.NET.Test.Sdk/) has custom Microsoft license.

It might seem like not a big deal, but we're using proxy server for all our NuGet packages. This server allows packages with MIT license, but blocks packages with custom license. Such blocked packages like Microsoft.NET.Test.Sdk need to be manually approved and it's quite tedious work for every new released package version.

Tl;dr changing the license of Microsoft.NET.Test.Sdk and related packages to MIT on NuGet, similarly to MSTest NuGets, would be highly appreciated!

**Comment by Evangelink** [comment]
Ping @pavelhorak

**Comment by pavelhorak** [comment]
This is because along with Microsoft.Net.Test.Sdk package we ship Microsoft.CodeCoverage package dependency which isn't MIT licensed.

**Comment by martinsuchan** [comment]
The [Microsoft.CodeCoverage](https://www.nuget.org/packages/Microsoft.CodeCoverage/) package links to [VSTest Source repository](https://github.com/microsoft/vstest) with MIT licence, so this answer does not make much sense.

**Comment by pavelhorak** [comment]
Code Coverage is closed source and we didn't have any public repo but we are about to create one for issue tracking, etc... Once done, we'll reference it from Nuget for [Microsoft.CodeCoverage](https://www.nuget.org/packages/Microsoft.CodeCoverage/) package.

**Comment by SimonCropp** [comment]
why does Microsoft.Net.Test.Sdk consume Microsoft.CodeCoverage? shouldnt that be an optional thing people pull in?

**Comment by Evangelink** [comment]
@jakubch1 for more details.

**Comment by jakubch1** [comment]
> why does Microsoft.Net.Test.Sdk consume Microsoft.CodeCoverage? shouldnt that be an optional thing people pull in?

I agree it should be optional. Unfortunately removing it would break many projects because `dotnet test --collect "Code Coverage"` will stop to work. I will try to make change to include `Microsoft.CodeCoverage` into default templates and after longer time we can remove dependency.

`Microsoft.Net.Test.Sdk ` is just some build props and targets + `Microsoft.NET.Test.Sdk.Program.cs` file that you can include into your project and get it behave like test project (you can directly reference `<PackageReference Include="Microsoft.TestPlatform.TestHost" Version="17.4.1" />`). It could be done as `Directory.Build.props` and `Directory.Build.targets` in folder with all your tests projects.
